[SPARK-23053][CORE] taskBinarySerialization and task partitions calculate in DagScheduler.submitMissingTasks should keep the same RDD checkpoint status

ivoson opened the pull request:
## What changes were proposed in this pull request?

When we run concurrent jobs using the same rdd which is marked to do checkpoint. If one job has finished running the job, and start the process of RDD.doCheckpoint, while another job is submitted, then submitStage and submitMissingTasks will be called. In [submitMissingTasks](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L961), will serialize taskBinaryBytes and calculate task partitions which are both affected by the status of checkpoint, if the former is calculated before doCheckpoint finished, while the latter is calculated after doCheckpoint finished, when run task, rdd.compute will be called, for some rdds with particular partition type such as [UnionRDD](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala) who will do partition type cast, will get a ClassCastException because the part params is actually a CheckpointRDDPartition.
This error occurs  because rdd.doCheckpoint occurs in the same thread that called sc.runJob, while the task serialization occurs in the DAGSchedulers event loop.

## How was this patch tested?

the exist uts and also add a test case in DAGScheduerSuite to show the exception case.

Commit:
submitMissingTasks should make sure the checkpoint status of stage.rdd is the same when calculate taskSerialization and task partitions

Change-Id: Ib9839ca552653343d264135c116742effa6feb60

AmplabJenkins commented:
Can one of the admins verify this patch?

ivoson commented:
@xuanyuanking could review this please?

xuanyuanking commented:
reopen this...

xuanyuanking commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -96,6 +98,22 @@ class MyRDD(
+  }
+
+  def compute(split: Partition, context: TaskContext): Iterator[Int] = {
+    parent.compute(split.asInstanceOf[WrappedPartition].partition, context)
```
I think this line is the key point for `WrppedPartition` and `WrappedRDD`, please give comments for explaining your intention.

ivoson replied:
thanks for the comment, i will work on this.

xuanyuanking commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -2399,6 +2417,93 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
+   * With this test case, just want to indicate that we should do taskSerialization and
+   * part calculate in submitMissingTasks with the same rdd checkpoint status.
+   */
+  test("task part misType with checkpoint rdd in concurrent execution scenes") {
```
maybe "SPARK-23053: avoid CastException in concurrent execution with checkpoint" better?

ivoson replied:
thanks for the suggest.

xuanyuanking commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -2399,6 +2417,93 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
+        val (taskRdd, taskFunc) = ser.deserialize[(RDD[Int], (TaskContext, Iterator[Int]) => Unit)](
+          ByteBuffer.wrap(taskBinaryBytes), Thread.currentThread.getContextClassLoader)
+        val part = rdd.partitions(0)
+        intercept[ClassCastException] {
```
I think this not a "test", this just a "reproduce" for the problem you want to fix. We should prove your code added in `DAGScheduler.scala` can fix that problem and with the original code base, a `ClassCastException` raised.

ivoson replied:
it is a reproduce case, i will fix this.

xuanyuanking commented:
test this please

xuanyuanking commented:
ok to test

Commit:
address comments

Change-Id: Id791079358808d3f6732f0c4fdb2703a75b0677e

xuanyuanking reviewed (approved):
LGTM, cc @zsxwing @jerryshao @gatorsmile @cloud-fan 
hi Shixiong, Saisai, Xiao and Wenchen. We found this streaming job problem in Baidu practical, Spark version 2.1. This triggered by streaming checkpoint workload and the fix patch mainly on core module. Please give a review. Thanks :)

xuanyuanking commented:
@ivoson Tengfei, please post the full stack trace of the `ClassCastException`.

ivoson commented:
This is the stack trace of the Exception.

```
java.lang.ClassCastException: org.apache.spark.rdd.CheckpointRDDPartition cannot be cast to org.apache.spark.streaming.rdd.MapWithStateRDDPartition
at org.apache.spark.streaming.rdd.MapWithStateRDD.compute(MapWithStateRDD.scala:152)
at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
at org.apache.spark.rdd.RDD$$anonfun$8.apply(RDD.scala:336)
at org.apache.spark.rdd.RDD$$anonfun$8.apply(RDD.scala:334)
at org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:957)
at org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:948)
at org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:888)
at org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:948)
at org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:694)
at org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:334)
at org.apache.spark.rdd.RDD.iterator(RDD.scala:285)
```

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

squito commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1016,15 +1016,23 @@ class DAGScheduler(
-          JavaUtils.bufferToArray(closureSerializer.serialize((stage.rdd, stage.func): AnyRef))
+      var taskBinaryBytes: Array[Byte] = null
+      // Add synchronized block to avoid rdd deserialized from taskBinaryBytes has diff checkpoint
+      // status with the rdd when create ShuffleMapTask or ResultTask.
```
I'd reword this a bit:

taskBinaryBytes and partitions are both effected by the checkpoint status.  We need this synchronization in case another concurrent job is checkpointing this RDD, so we get a consistent view of both variables.

ivoson replied:
thanks for the advise.

squito commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
+    // set checkpointDir.
+    val tempDir = Utils.createTempDir()
+    val checkpointDir = File.createTempFile("temp", "", tempDir)
+    checkpointDir.delete()
```
why do you make a tempfile for the checkpoint dir and then delete it?  why not just  `checkpointDir = new File(tempDir, "checkpointing")`?  Or even just `checkpointDir = Utils.createTempDir()`?

(CheckpointSuite does this so it can call `sc.setCheckpointDir`, but you're not doing that here)

ivoson replied:
check the code again and yes ```checkpointDir = Utils.createTempDir()``` is enough for this case, will fix this.

squito commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
+
+    val checkpointRunnable = new Runnable {
+      override def run() = {
+        // Simply simulate what RDD.doCheckpoint() do here.
```
I'd remove "simply" here and elsewhere in comments.  Also "do" -> "does"

ivoson replied:
will fix this.

squito commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
+        // serialization can start.
+        semaphore1.release()
+        // Wait until taskBinary serialization finished in submitMissingTasksThread.
+        semaphore2.acquire()
```
this would be a bit easier to follow if you rename your semaphores a bit.

`semaphore1` -> `doCheckpointStarted`
`semaphore2` -> `taskBinaryBytesFinished`

ivoson replied:
thanks for the advise. will fix this.

squito commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
+          checkpointData.cpState = CheckpointState.Checkpointed
+          rdd.markCheckpointed()
+        }
+        semaphore1.release()
```
and then this would be another semaphore `checkpointStateUpdated`

ivoson replied:
thanks for the advise

squito commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
+        // Wait until checkpoint status changed to Checkpointed in checkpointThread.
+        semaphore1.acquire()
+
+        // Part calculated with rdd checkpoint already finished.
```
I'd add a comment above this:

Now we're done simulating the interleaving that might happen within the scheduler -- we'll check to make sure the final state is OK by simulating a couple steps that normally happen on the executor.

ivoson replied:
thanks for the advise, it is really helpful for understanding, will update this.

squito commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
+        val taskContext = mock(classOf[TaskContext])
+        doNothing().when(taskContext).killTaskIfInterrupted()
+
+        // ClassCastException is expected with errPart.
```
I think this is a bit easier to follow if you say

Make sure our test case is setup correctly -- we expect a ClassCastException here if we use the `rdd.partitions` *after* checkpointing was done, but our binary bytes is from before it finished.

ivoson replied:
thanks for the advise, it is really helpful for understanding, will update this.

squito commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
+    submitMissingTasksThread.start()
+    submitMissingTasksThread.join()
+
+    Utils.deleteRecursively(tempDir)
```
this should be done in a `finally`.

ivoson replied:
will fix this.

squito reviewed (changes requested):
Thanks a lot for finding & working on this @ivoson.  Overall this fix looks great, I just have some small comments for clarity.

Two other small asks:

1) Both in the PR description and the JIRA, can you add that this error occurs because `rdd.doCheckpoint` occurs in the same thread that called `sc.runJob`, while the task serialization occurs in the DAGSchedulers event loop?

2) I saw you put the stack trace in a PR comment, can you also put it on the JIRA?  I think that is easier to find for most users.

Commit:
address comments

Change-Id: Id49321f8761d03159fd48b6f607084f2cc6fa4ff

ivoson commented:
@squito Hi Rashid, thanks for your review and advise. The PR description and JIRA have been updated , and also put the stack trace on the JIRA. 
The last commit addresses the comments you left. Thanks again for the advise.

squito commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -2399,6 +2424,121 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
+   * With this test case, just want to indicate that we should do taskSerialization and
+   * part calculate in submitMissingTasks with the same rdd checkpoint status.
+   */
+  test("SPARK-23053: avoid ClassCastException in concurrent execution with checkpoint") {
```
hi @ivoson  -- I'm really sorry but I only just realized that this "test" is really just a repro, and it passes both before and after the actual code changes, since you've replicated the internal logic we're fixing.  As such, I don't think its actually useful as a test case -- perhaps it should get added to the jira as a repro.

I appreciate the work that went into writing this as it helped make the issue clear to me.  I am not sure if there is a good way to test this.  If we can't come up with anything, we should just commit your actual fix, but give me a day or two to think about it ...

ivoson replied:
@squito thanks for reply. I understand this, technically it may not be a UT case, just simulate the scene with exception. I also wonder if there is a good way to test this.

squito replied:
hi @ivoson -- I haven't come up with a better way to test this, so I think for now you should

(1) change the PR to *only* include the changes to the DAGScheduler (also undo the `protected[spark]` changes elsewhere)
(2) put this repro on the jira as its a pretty good for showing whats going on.

if we come up with a way to test it, we can always do that later on.

thanks and sorry for the back and forth

ivoson replied:
hi @squito , it's fine. The pr and jira have been updated. Thanks for your patient and review.

Commit:
remove the test case and undo the protected[spark] changes

Change-Id: I6c308d5953a243e30dae87e8109e25d5df5a3f91

squito reviewed (approved):
lgtm

squito commented:
merged to master / 2.3 / 2.2

I hit a merge conflict trying to merge to 2.1 -- feel free to open another PR for that version.

ivoson commented:
thank you for reviewing this @squito
